### PR TITLE
feat(platform)!: add owner keys to identities, fixed verification of use of owner keys

### PR DESCRIPTION
**Pull request by QuantumExplorer**
## Issue being fixed or feature implemented
There were 2 big issues, Masternodes didn't actually have their owner keys set. And owner keys could sign withdrawals that were sent to a script payout address.

## What was done?
Introduced an Owner Key Purpose, when we change to protocol version 4, all active Masternodes will set this owner key purpose. All new masternodes/evonodes will have this set.

Normal identities can only withdraw with their transfer keys.
Masternode identities can withdraw with transfer keys anywhere. Or with the owner key, but only to their payout address.

## How Has This Been Tested?
Added a lot of tests.

## Breaking Changes
While these are breaking changes, they will only happen in Protocol V4.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

## Summary by CodeRabbit

- **New Features**
  - Introduced a new error variant for handling withdrawal output scripts.
  - Added new methods for generating keys related to masternodes, enhancing key management capabilities.
  - Expanded the `Purpose` enum to include a new `OWNER` variant for ownership verification.
  - Updated error handling framework with new error variants for specific withdrawal scenarios.
  - Added a method for retrieving an identity public key based on an owner's key and key ID.

- **Bug Fixes**
  - Simplified import statements to improve code clarity without affecting functionality.
  - Updated test case descriptions to reflect changes in expected behavior related to key purposes.

- **Documentation**
  - Updated documentation to reflect new error handling capabilities and key generation methods.

### CODE REVIEW

File: packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_masternode_identities/create_owner_identity/v1/mod.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+                masternode.state.payout_address,
+                0,
+                platform_version,
+            )?,
+            Self::get_owner_identity_owner_key(
+                masternode.state.owner_address,
+                1,
+                platform_version,
+            )?,
+        ]);
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_masternode_identities/create_owner_identity/v1/mod.rs:30]: add_public_keys can not fail

File: packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_masternode_identities/update_owner_withdrawal_address/v1/mod.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            let new_owner_withdrawal_key = Self::get_owner_identity_withdrawal_key(
+                new_withdrawal_address,
+                last_key_id + 1,
+                platform_version,
+            )?;
+
+            tracing::trace!(
+                identity_id = ?owner_identifier,
+                withdrawal_key = ?new_owner_withdrawal_key,
+                method = "update_owner_withdrawal_address_v0",
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_masternode_identities/update_owner_withdrawal_address/v1/mod.rs:123]: thank you

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs [hunk 3]
```diff
@@ -97,8 +100,24 @@ Your software version: {}, latest supported protocol version: {}."#,
                 // Set current protocol version to the block platform state
                 block_platform_state
                     .set_current_protocol_version_in_consensus(next_protocol_version);
+
+                let last_block_time_ms = platform_state.last_committed_block_time_ms();
+
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs:105]: no, this is fine

File: packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs [hunk 4]
```diff
@@ -97,8 +100,24 @@
...
+                let block_info = block_state_info.to_block_info(
+                    Epoch::new(epoch_info.current_epoch_index())
+                        .expect("current epoch index should be in range"),
+                );
+
                 // This is for events like adding stuff to the root tree, or making structural changes/fixes
                 self.perform_events_on_first_block_of_protocol_change(
+                    platform_state,
+                    &block_info,
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/engine/run_block_proposal/mod.rs:120]: no this is fine

File: packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_masternode_identities/create_owner_identity/v1/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_masternode_identities/create_owner_identity/v1/mod.rs:30]: No, there is no result

File: packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_masternode_identities/get_owner_identity_owner_key/mod.rs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+impl<C> Platform<C> {
+    /// Retrieves an identity public key using the provided owner key and key ID.
+    ///
+    /// This function derives the identity public key and delegates to a version-specific method depending on the platform version.
+    ///
+    /// # Arguments
+    ///
+    /// * payout_address - The payout address for the owner.
+    /// * key_id - The KeyID for the identity public key.
+    /// * platform_version - The version of the platform to determine which method to delegate to.
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/platform_events/core_based_updates/update_masternode_identities/get_owner_identity_owner_key/mod.rs:20]: fixed.

File: packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/identity_credit_withdrawal/signature_purpose_matches_requirements/v0/mod.rs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                    WithdrawalOutputScriptNotAllowedWhenSigningWithOwnerKeyError::new(
+                        output_script.clone(),
+                        signing_key.id(),
+                    ),
+                );
+            }
+        }
+
+        Ok(result)
+    }
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/identity_credit_withdrawal/signature_purpose_matches_requirements/v0/mod.rs:48]: no this is fine

File: packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/identity_credit_withdrawal/signature_purpose_matches_requirements/v0/mod.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        _platform_version: &PlatformVersion,
+    ) -> Result<SimpleConsensusValidationResult, Error> {
+        let mut result = SimpleConsensusValidationResult::default();
+
+        if let Some(output_script) = self.output_script() {
+            let Some((_, signing_key)) = identity.loaded_public_keys.first_key_value() else {
+                return Err(Error::Execution(ExecutionError::CorruptedCodeExecution(
+                    "we should have a loaded key at this point",
+                )));
+            };
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/identity_credit_withdrawal/signature_purpose_matches_requirements/v0/mod.rs:35]: good point

File: packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/mod.rs [hunk 8]
```diff
@@ -512,6 +512,111 @@
...
+        let mut platform_state = platform.state.load().clone().deref().clone();
+
+        let pro_tx_hash = ProTxHash::from_byte_array(pro_tx_hash_bytes);
+
+        let random_ip = Ipv4Addr::new(
+            rng.gen_range(0..255),
+            rng.gen_range(0..255),
+            rng.gen_range(0..255),
+            rng.gen_range(0..255),
+        );
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply packages/rs-drive-abci/src/execution/validation/state_transition/state_transitions/mod.rs:586]: this is fine

**Review by shumkov — APPROVED** [review]

**Pull request merged**